### PR TITLE
Support ProjectReference

### LINKED ISSUE #568: ProjectReference causes nccs to crash
If the project has a 'project reference' in it, nccs crashes:

```
D:\Sources\neo-devpack-dotnet\src\Neo.Compiler.CSharp\bin\Debug\net5.0>**nccs  D:\Sources\Hello2\Hello2Contract.csproj**
  Determining projects to restore...
  All projects are up-to-date for restore.
Unhandled exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Neo.Compiler.CompilationContext.CompileProject(String csproj, Options options) in D:\Sources\neo-devpack-dotnet\src\Neo.Compiler.CSharp\CompilationContext.cs:line 160
...

```
image

Since it is looking for the entry "[libraries][name][files]" but there is no [files] entry for "project" type:

image

Not sure if this story (e.g., adding project references to smart contract project) is supported or not. At least, we need better error report if not.

### PR DESCRIPTION
**Pull request by erikzhang**
Close #568

### CODE REVIEW

**Comment by erikzhang** [comment]
After some research, I still can't find the reason why it doesn't work.

**Comment by wschae** [comment]
This PR works for me. No more crash for project reference and the generated nef looks good to me. I could deploy it to neo-express and see it working (e.g., invoked and see the result). Somehow, I couldn't run the debugger though.

Can you elaborate what was not working?

**Comment by superboyiii** [comment]
It also works for me.
image

**Comment by superboyiii** [comment]
> After some research, I still can't find the reason why it doesn't work.

Could you paste your project here?

**Comment by erikzhang** [comment]
[Test.zip](https://github.com/neo-project/neo-devpack-dotnet/files/6316033/Test.zip)

@superboyiii

**Comment by superboyiii** [comment]
> [Test.zip](https://github.com/neo-project/neo-devpack-dotnet/files/6316033/Test.zip)
> 
> @superboyiii
> 
> ```
>   正在确定要还原的项目…
>   已还原 C:\Users\erik\source\repos\nogit\Test\TestProjectReference\TestProjectReference.csproj (用时 74 ms)。
>   已还原 C:\Users\erik\source\repos\nogit\Test\TestContract\TestContract.csproj (用时 117 ms)。
>   正在确定要还原的项目…
>   所有项目均是最新的，无法还原。
> C:\Users\erik\source\repos\nogit\Test\TestContract\TestContract.cs(3,30): error CS0246: 未能找到类型或命名空间名“Framework”(是否缺少 using 指令或程序集引用?)
> Compilation failed.
> ```

The same as mine.
```
E:\Master-smoking\2021-04-15\Test\TestContract\TestContract.cs(3,38): error CS0246: The type or namespace name 'Framework' could not be found (are you missing a using directive or an assembly reference?)
```

**Comment by erikzhang** [comment]
I don't know why.

**Comment by wschae** [comment]
In Test.zip, I saw this error:

```
D:\Sources\neo-devpack-dotnet\src\Neo.Compiler.CSharp\bin\Debug\net5.0>nccs  C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj
  Determining projects to restore...
C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj : error NU1102: Unable to find package Neo.SmartContract.Framework with version (>= 3.0.0-rc1-04)
C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj : error NU1102:   - Found 16 version(s) in nuget.org [ Nearest version: 3.0.0-rc1 ]
C:\Users\wochae.\Downloads\Test\TestContract\TestContract.csproj : error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
  Failed to restore C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj (in 788 ms).
  1 of 2 projects are up-to-date for restore.
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Users\wochae\Downloads\Test\TestContract\TestContract.cs(3,30): error CS0246: The type or namespace name 'Framework' could not be found (are you missing a using directive or an assembly reference?)
```

Since I didn't have "3.0.0-rc1-04", I updated "**3.0.0-rc1-04**" to "**3.0.0-rc1**" and the issue disappeared. I may miss something but hope this might give some idea what went wrong.

**Comment by erikzhang** [comment]
@wschae I packed this pr to a nuget package locally, and changed the version to 3.0.0-rc1-04.

**Comment by superboyiii** [comment]
> In Test.zip, I saw this error:
> 
> ```
> D:\Sources\neo-devpack-dotnet\src\Neo.Compiler.CSharp\bin\Debug\net5.0>nccs  C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj
>   Determining projects to restore...
> C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj : error NU1102: Unable to find package Neo.SmartContract.Framework with version (>= 3.0.0-rc1-04)
> C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj : error NU1102:   - Found 16 version(s) in nuget.org [ Nearest version: 3.0.0-rc1 ]
> C:\Users\wochae.\Downloads\Test\TestContract\TestContract.csproj : error NU1102:   - Found 0 version(s) in Microsoft Visual Studio Offline Packages
>   Failed to restore C:\Users\wochae\Downloads\Test\TestContract\TestContract.csproj (in 788 ms).
>   1 of 2 projects are up-to-date for restore.
>   Determining projects to restore...
>   All projects are up-to-date for restore.
> C:\Users\wochae\Downloads\Test\TestContract\TestContract.cs(3,30): error CS0246: The type or namespace name 'Framework' could not be found (are you missing a using directive or an assembly reference?)
> ```
> 
> Since I didn't have "3.0.0-rc1-04", I updated "**3.0.0-rc1-04**" to "**3.0.0-rc1**" in 'TestContract.csproj' and the issue disappeared. I may miss something but hope this might give some idea what went wrong.

You could try modify it to Neo.SmartContract.Framework v3.0.0-CI00300.
```
  <ItemGroup>
    <PackageReference Include="Neo.SmartContract.Framework" Version="3.0.0-CI00300" />
  </ItemGroup>
```

File: src/Neo.Compiler.CSharp/CompilationContext.cs [hunk 1]
```diff
@@ -134,53 +134,73 @@ public static CompilationContext CompileSources(string[] sourceFiles, Options op
...
+                                if (file.EndsWith("_._")) continue;
+                                string path = Path.Combine(packagesPath, name, file);
+                                if (!File.Exists(path)) continue;
+                                references.Add(MetadataReference.CreateFromFile(path));
+                            }
+                        }
+                        else
+                        {
+                            IEnumerable<SyntaxTree> st = files.OrderBy(p => p).Select(p => Path.Combine(packagesPath, name, p)).Select(p => CSharpSyntaxTree.ParseText(File.ReadAllText(p), path: p));
+                            CSharpCompilation cr = CSharpCompilation.Create(null, st, commonReferences, options);
```
Comment (Ashuaidehao) [inline src/Neo.Compiler.CSharp/CompilationContext.cs:181]: Maybe assembly should has an identity name for use it.

Comment (erikzhang) [reply src/Neo.Compiler.CSharp/CompilationContext.cs]: That's it! Fixed!

**Review by wschae — APPROVED** [review]
Looks good to me. Thanks for fixing it 👍

**Pull request merged**